README.md
# NewsAPIClient

[![CI Status](http://img.shields.io/travis/andreamarcolin/NewsAPIClient.svg?style=flat)](https://travis-ci.org/andreamarcolin/NewsAPIClient)
[![Version](https://img.shields.io/cocoapods/v/NewsAPIClient.svg?style=flat)](http://cocoapods.org/pods/NewsAPIClient)
[![License](https://img.shields.io/cocoapods/l/NewsAPIClient.svg?style=flat)](http://cocoapods.org/pods/NewsAPIClient)
[![Platform](https://img.shields.io/cocoapods/p/NewsAPIClient.svg?style=flat)](http://cocoapods.org/pods/NewsAPIClient)

## Requirements

iOS 8.0+ 
Swift 3.0+

## Installation

NewsAPIClient is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "NewsAPIClient"
```

## Usage

First, create a client with your NewsAPI.org API Key

```swift
let client = NewsAPIClient(apiKey: "<YOUR_API_KEY>")
```

The client provides two methods to interact with the two API endpoints: "/sources" and "/articles".
N.B. all operations are performed on the main thread. It is highly reccomended to dispatch each request to a background queue, and pass a completion handler which eventually performs UI updates on the main thread, accordingly with your needs. A general example can be found at http://stackoverflow.com/questions/24056205/how-to-use-background-thread-in-swift

### Sources

Send a request to the "/sources" endpoint to get all sources

```swift
client.getSources { (sources, error) in
    guard let sources = sources else {
        print(error!)
        return
    }
    print(sources)
}
```

Optionally filter the sources by category, language or country

```swift
client.getSources(category: "business",
                  language: "en",
                  country: "gb")
{ (sources, error) in

    guard let sources = sources else {
        print(error!)
        return
    }

    print(sources)
}
```

### Articles

Normally, chain the sources request with a request to the "/sources" endpoint to get articles from a specific source (sortBy is optional)

```swift
client.getSources { (sources, error) in
    guard let sources = sources else {
        print(error!)
        return
    }

    client.getArticles(sourceId: sources[0].id,
                       sortBy: sources[0].availableSortBys["latest"]) // if "latest" is not available for this source, defaults to "top" 
    { (articles, error) in

        guard let articles = articles else {
            print(error!)
            return
        }

        print(articles)
    }
}
```

Otherwise, if you already have a source object (i.e. previously obtained by the source method) just pass it to the getArticles() method to obtain articles for the given source
Remember, sortby is optional, so in this case we provide an example without passing a sortBy.

```swift
client.getArticles(source: source) { (articles, error) in
    guard let articles = articles else {
        print(error!)
        return
    }

    print(articles)
}
```

## Author

Andrea Marcolin, andreamarcolin@hotmail.it

## License

NewsAPIClient is available under the MIT license. See the LICENSE file for more info.


Example/CHANGELOG.md
## 0.2.6 (2017-03-21)
* Better error handling for network errors

## 0.2.5 (2017-03-19)
* Added support for etherogeneous date formats for article's "published at" date

## 0.2.4 (2017-03-08)
* Fixed typo introduced in previous release

## 0.2.3 (2017-03-07)
* Bug fixes on articles endpoint

## 0.2.2 (2017-02-11)
* Bug fix on sources endpoint

## 0.2.1 (2017-01-03)

* Updated README and docs
* Added CHANGELOG

## 0.2.0 (2017-01-03)

* Simplified Article and Source's structure for easier integration

## 0.1.0 (2017-01-01)

* Initial release


Example/Tests/Tests.swift
import UIKit
import XCTest
@testable import NewsAPIClient

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    func testSourceInitialization() {
        
        let id = "bbc-news"
        let name = "BBC News"
        let description = "Use BBC News for up-to-the-minute news, breaking news, video, audio and feature stories. BBC News provides trusted World and UK news as well as local and regional perspectives. Also entertainment, business, science, technology and health news."
        let url = "http://www.bbc.co.uk/news"
        let category = "general"
        let language = "en"
        let country = "gb"
        let logoUrls = ["small" : "http://i.newsapi.org/bbc-news-s.png",
                        "medium" : "http://i.newsapi.org/bbc-news-m.png",
                        "large" : "http://i.newsapi.org/bbc-news-l.png"]
        let availableSortBys = ["top"]
        
        let source = NewsAPIClient.Source(id: id,
                                          name: name,
                                          description: description,
                                          url: url,
                                          category: category,
                                          language: language,
                                          country: country,
                                          logoUrls: logoUrls,
                                          availableSortBys: availableSortBys)
        
        XCTAssertEqual(source.id, "bbc-news")
        XCTAssertEqual(source.name, "BBC News")
        XCTAssertEqual(source.description, "Use BBC News for up-to-the-minute news, breaking news, video, audio and feature stories. BBC News provides trusted World and UK news as well as local and regional perspectives. Also entertainment, business, science, technology and health news.")
        XCTAssertEqual(source.url, "http://www.bbc.co.uk/news")
        XCTAssertEqual(source.category, "general")
        XCTAssertEqual(source.language, "en")
        XCTAssertEqual(source.country, "gb")
        XCTAssertEqual(source.logoUrls, ["small" : "http://i.newsapi.org/bbc-news-s.png",
                                         "medium" : "http://i.newsapi.org/bbc-news-m.png",
                                         "large" : "http://i.newsapi.org/bbc-news-l.png"])
        XCTAssertEqual(source.availableSortBys, ["top" : "top"])
    }
    
    func testArticleInitialization() {
        
        let sourceId = "the-next-web"
        let author = "Arno Nijhof"
        let title = "The hunt for the fastest growing company of Europe starts now"
        let description = "Tech5 is a competition in which TNW and Adyen search for the fastest growing tech companies to celebrate their growth. Companies can apply from 1 January."
        let url = "http://thenextweb.com/insider/2017/01/03/the-hunt-for-the-fastest-growing-company-of-europe-starts-now/"
        let urlToImage = "https://cdn3.tnwcdn.com/wp-content/blogs.dir/1/files/2016/12/Banner-Tech5.png"
        let publishedAt = "2017-01-03T08:49:12Z"
        
        do {
            let article = try NewsAPIClient.Article(sourceId: sourceId,
                                                author: author,
                                                title: title,
                                                description: description,
                                                url: url,
                                                urlToImage: urlToImage,
                                                publishedAt: publishedAt)
            XCTAssertEqual(article.sourceId, "the-next-web")
            XCTAssertEqual(article.author, "Arno Nijhof")
            XCTAssertEqual(article.title, "The hunt for the fastest growing company of Europe starts now")
            XCTAssertEqual(article.description, "Tech5 is a competition in which TNW and Adyen search for the fastest growing tech companies to celebrate their growth. Companies can apply from 1 January.")
            XCTAssertEqual(article.url, "http://thenextweb.com/insider/2017/01/03/the-hunt-for-the-fastest-growing-company-of-europe-starts-now/")
            XCTAssertEqual(article.urlToImage, "https://cdn3.tnwcdn.com/wp-content/blogs.dir/1/files/2016/12/Banner-Tech5.png")
            
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
            XCTAssertEqual(article.publishedAt, dateFormatter.date(from: "2017-01-03T08:49:12Z"))
            
        } catch {
            XCTFail("Unexpected error on Article initialization")
        }
    }
    
    func testArticleInitializationFail() {
        let sourceId = "the-next-web"
        let author = "Arno Nijhof"
        let title = "The hunt for the fastest growing company of Europe starts now"
        let description = "Tech5 is a competition in which TNW and Adyen search for the fastest growing tech companies to celebrate their growth. Companies can apply from 1 January."
        let url = "http://thenextweb.com/insider/2017/01/03/the-hunt-for-the-fastest-growing-company-of-europe-starts-now/"
        let urlToImage = "https://cdn3.tnwcdn.com/wp-content/blogs.dir/1/files/2016/12/Banner-Tech5.png"
        let publishedAt = "badDate"
        
        XCTAssertThrowsError(try NewsAPIClient.Article(sourceId: sourceId,
                                                   author: author,
                                                   title: title,
                                                   description: description,
                                                   url: url,
                                                   urlToImage: urlToImage,
                                                   publishedAt: publishedAt))
    }
    
    func testWrongAPIKey() {
        
        let badApiKey = "badApiKey"
        let client = NewsAPIClient(apiKey: badApiKey)
        
        client.getSources { (sources, error) in
            XCTAssertNil(sources)
            XCTAssertNotNil(error)
        }
        
    }
    
    // FIXME: tests not working on Travis CI 
    // cause: "fatal error: unexpectedly found nil while unwrapping optional value"
    /*
    func testGetSources() {
        
        let apiKey = "5f91ae4f530f453b88c331b7a33a331e"
        let client = NewsAPIClient(apiKey: apiKey)
        
        self.measure {
            client.getSources { (sources, error) in
                XCTAssertNil(error)
                XCTAssertNotNil(sources)
            }
        }
    }
 
    func testGetArticles() {
        let apiKey = "5f91ae4f530f453b88c331b7a33a331e"
        let client = NewsAPIClient(apiKey: apiKey)
        
        self.measure {
            client.getSources { (sources, error) in
                XCTAssertNil(error)
                XCTAssertNotNil(sources)
                
                client.getArticles(sourceId: sources![0].id,
                                   sortBy: sources![0].availableSortBys["latest"])
                { (articles, error) in
                    XCTAssertNil(error)
                    XCTAssertNotNil(articles)
                }
            }
        }
    }
    */
}


NewsAPIClient/Classes/Source.swift
//
//  Source.swift
//  NewsAPIClient
//
//  Created by Andrea Marcolin (Twitter @andreamarcolin5) on 14/12/16.
//  Some rights reserved: http://opensource.org/licenses/MIT
//

import Foundation

public extension NewsAPIClient {
    
    /// A struct which represents a Source entity from NewsAPI.org APIs
    public struct Source {
        
        public let id: String
        public let name: String
        public let description: String
        public let url: String
        public let category: String
        public let language: String
        public let country: String
        public let logoUrls: [String : String]
        public let availableSortBys: [String : String]
        
        /// The init method for Source
        ///
        /// - Parameters:
        ///   - id: the unique identifier for the news source. This is needed when querying the /articles endpoint to retrieve article metadata
        ///   - name: the display-friendly name of the news source
        ///   - description: a brief description of the news source and what area they specialize in
        ///   - url: the base URL or homepage of the source
        ///   - category: the topic category that the source focuses on
        ///   - language: the 2-letter ISO-639-1 code for the language that the source is written in
        ///   - country: the 2-letter ISO 3166-1 code of the country that the source mainly focuses on
        ///   - logoUrls: an array containing URLs to the source's logo, in different sizes
        ///   - availableSortBys: the available headline lists for the news source
        /// - Throws: an InitializationError if somthing fails during properties initialization (eg. invalid URLs or unknown languages, categories and countries)
        init(id: String, name: String, description: String, url: String, category: String, language: String, country: String, logoUrls: [String : String], availableSortBys: [String]) {
            
            // Assign the parsed objects to struct properties
            self.id = id
            self.name = name
            self.description = description
            self.url = url
            self.category = category
            self.language = language
            self.country = country
            self.logoUrls = logoUrls
            
            // Build a dictionary from array values with same key as value, so we can take advantage of access by String
            var parsedAvailableSortBys = [String : String]()
            for sortBy in availableSortBys {
                parsedAvailableSortBys[sortBy] = sortBy
            }
            self.availableSortBys = parsedAvailableSortBys
        }
    } 
}



NewsAPIClient/Classes/Article.swift
//
//  Article.swift
//  NewsAPIClient
//
//  Created by Andrea Marcolin (Twitter @andreamarcolin5) on 14/12/16.
//  Some rights reserved: http://opensource.org/licenses/MIT
//

import Foundation

public extension NewsAPIClient {
    
    /// A struct which represents an Article entity from NewsAPI.org APIs
    public struct Article {
        
        public let sourceId: String
        public let author: String
        public let title: String
        public let description: String
        public let url: String
        public let urlToImage: String
        public let publishedAt: Date
        
        /// The init method for Article
        ///
        /// - Parameters:
        ///   - author: the author of the article
        ///   - title: the headline or title of the article.
        ///   - description: a description or preface for the article (eg. the first part of the body)
        ///   - url: the direct URL to the content page of the article
        ///   - urlToImage: the URL to a relevant image for the article
        ///   - publishedAt: the best attempt at finding a date for the article, in UTC (+0). Format for the date is "yyyy-MM-dd'T'HH:mm:ss'Z'" (eg. "2016-12-16T09:56:38Z")
        /// - Throws: an InitializationError if somthing fails during properties initialization (eg. invalid URLs or date format)
        init(sourceId: String, author: String, title: String, description: String, url: String, urlToImage: String, publishedAt: String) throws {
            
            // Parse the publishedAt date as Date, guarding for possible unrecognized values (if it is the case, throw an error and exit early)
            let dateFormatter = DateFormatter()
            
            // Try with different date formats while the suitable one is found
            for format in Constants.Utilities.dateFormats {
                dateFormatter.dateFormat = format
                if let date = dateFormatter.date(from: publishedAt) {
                    // Found a suitable date format
                    // Assign the parsed objects to struct properties
                    self.sourceId = sourceId
                    self.author = author
                    self.title = title
                    self.description = description
                    self.url = url
                    self.urlToImage = urlToImage
                    self.publishedAt = date
                    return
                }
            }
            // If we came here, no suitable date format was found, return an errors
            throw InitializationError.invalidPublishedAtDate(invalidPublishedAtDate: publishedAt)
        }
        
        /// The possible errors which can arise on Article struct initilization
        ///
        /// - invalidPublishedAtDate: when a date is not in the correct format (cannot be parsed from String to Date with the format "yyyy-MM-dd'T'HH:mm:ss'Z'")
        public enum InitializationError: Error {
            case invalidPublishedAtDate(invalidPublishedAtDate: String)
        }
    }
}


NewsAPIClient/Classes/NewsAPIClient.swift
//
//  NewsAPIClient.swift
//  NewsAPIClient
//
//  Created by Andrea Marcolin (Twitter @andreamarcolin5) on 14/12/16.
//  Some rights reserved: http://opensource.org/licenses/MIT
//

import Foundation

public class NewsAPIClient {
    
    fileprivate let apiKey: String
    
    /// A client to interact with NewsAPI.org REST APIs
    ///
    /// - Parameter apiKey: the API Key which is needed to sign requests
    public init(apiKey: String) {
        self.apiKey = apiKey
    }
    
    // MARK: - NewsAPI.org API methods
    
    /// Gets all articles for a NewsAPI.org source, optionally ordered by a given criteria
    ///
    /// - Parameters:
    ///   - source: the source id to get all Articles from
    ///   - sortBy: an optional criteria to order the articles by. Since not every Source can be ordered by all the possible criterias, choose between those which are stored in availableSortBys property of the source. Check https://newsapi.org for further details.
    ///   - completionHandler: a closure to be executed after the method completes, to which an array of Article structs (or an error) is passed
    public func getArticles(sourceId: String, sortBy: String? = "top", completionHandler: @escaping ([Article]?, Error?) -> Void) {
        
        // Build a dictionary of parameters for the request
        let parameters = [Constants.JSONKeys.source : sourceId,
                          Constants.JSONKeys.sortBy : sortBy!,
                          Constants.JSONKeys.apiKey : apiKey]
        
        taskForGetMethod(endpoint: Constants.Endpoints.articles, parameters: parameters) { (result, error) in
            
            guard (error == nil) else {
                completionHandler(nil, ResponseError.networkError(message: error!.localizedDescription))
                return
            }
            
            // Check if the status is "ok"
            guard let status = result![Constants.JSONKeys.status] as? String, status == Constants.Status.ok else {
                completionHandler(nil, ResponseError.apiError(message: result![Constants.JSONKeys.message] as? String))
                return
            }
            
            // Check if Source Id is present
            guard let sourceId = result![Constants.JSONKeys.source] as? String else {
                completionHandler(nil, ResponseError.noSource)
                return
            }
            
            // Check if articles are present
            guard let articles = result![Constants.JSONKeys.articles] as? [[String:AnyObject]] else {
                completionHandler(nil, ResponseError.noArticles)
                return
            }
            
            // Check for empty articles
            guard !articles.isEmpty else {
                completionHandler(nil, ResponseError.emptyArticles)
                return
            }
            
            // Build the result
            var parsedArticles = [Article]()
            for article in articles {
                
                // Allow empty Author or urlToImage (non-essential fields)
                // This is done because of the high number of articles not containing an author
                let author: String = (article[Constants.JSONKeys.author] as? String) ?? ""
                let urlToImage: String = (article[Constants.JSONKeys.urlToImage] as? String) ?? ""
                
                // Check if everything else is present and in the correct format
                guard let title = article[Constants.JSONKeys.title] as? String,
                    let description = article[Constants.JSONKeys.description] as? String,
                    let url = article[Constants.JSONKeys.url] as? String,
                    let publishedAt = article[Constants.JSONKeys.publishedAt] as? String else {
                    
                    // If one of these fields is not a String, skip the article
                    break
                }
                
                // Initialize a Source object with this source's data
                do {
                    let parsedArticle = try Article(sourceId: sourceId,
                                                    author: author,
                                                    title: title,
                                                    description: description,
                                                    url: url,
                                                    urlToImage: urlToImage,
                                                    publishedAt: publishedAt)
                    parsedArticles.append(parsedArticle)
                } catch let error {
                    completionHandler(nil, error)
                    return
                }
            }
            
            // Everything went good, return the parsed sources with no error
            completionHandler(parsedArticles, nil)
        }
    }
    
    /// Gets all news sources supported by NewsAPI.org
    ///
    /// - Parameters:
    ///   - category: the category you would like to get sources for (optional, check supported categories on https://newsapi.org)
    ///   - language: the language you would like to get sources for (optional, check supported languages on https://newsapi.org)
    ///   - country: the country you would like to get sources for (optional, check supported countries on https://newsapi.org)
    ///   - completionHandler: a closure to be executed after the method completes, to which an array of Source structs (or an error) is passed
    public func getSources(category: String? = nil, language: String? = nil, country: String? = nil, completionHandler: @escaping ([Source]?, Error?) -> Void) {
        
        // Build a dictionary of parameters for the request
        var parameters = [String : String]()
        
        if let category = category {
            parameters[Constants.JSONKeys.category] = category
        }
        if let language = language {
            parameters[Constants.JSONKeys.language] = language
        }
        if let country = country {
            parameters[Constants.JSONKeys.country] = country
        }
        
        taskForGetMethod(endpoint: Constants.Endpoints.sources) { (result, error) in
            
            // If there was an error on the URL session task, pass the error through completion handler
            guard (error == nil) else {
                completionHandler(nil, ResponseError.networkError(message: error!.localizedDescription))
                return
            }
            
            // Check if the status is "ok"
            guard let status = result![Constants.JSONKeys.status] as? String, status == Constants.Status.ok else {
                completionHandler(nil, ResponseError.apiError(message: result![Constants.JSONKeys.message] as? String))
                return
            }
            
            // Check if sources are present
            guard let sources = result![Constants.JSONKeys.sources] as? [[String:AnyObject]] else {
                completionHandler(nil, ResponseError.noSources)
                return
            }
            
            // Check for empty sources
            guard !sources.isEmpty else {
                completionHandler(nil, ResponseError.emptySources)
                return
            }
            
            // Build the result
            var parsedSources = [Source]()
            for source in sources {
                
                // Check if everything is present and in the correct format
                guard let id = source[Constants.JSONKeys.id] as? String,
                    let name = source[Constants.JSONKeys.name] as? String,
                    let description = source[Constants.JSONKeys.description] as? String,
                    let url = source[Constants.JSONKeys.url] as? String,
                    let category = source[Constants.JSONKeys.category] as? String,
                    let language = source[Constants.JSONKeys.language] as? String,
                    let country = source[Constants.JSONKeys.country] as? String,
                    let logoUrls = source[Constants.JSONKeys.urlToLogos] as? [String : String],
                    let availableSortBys = source[Constants.JSONKeys.sortBysAvailable] as? [String]
                else {
                    completionHandler(nil, ResponseError.wrongSourceFormat)
                    return
                }
                
                // Initialize a Source object with this source's data
                parsedSources.append(
                    Source(id: id,
                           name: name,
                           description: description,
                           url: url,
                           category: category,
                           language: language,
                           country: country,
                           logoUrls: logoUrls,
                           availableSortBys: availableSortBys)
                )
            }
            
            // Everything went good, return the parsed sources with no error
            completionHandler(parsedSources, nil)
        }
    }
    
    // MARK: - Helpers
    
    /// Handles the HTTP Request building, execution, parsing of the results and error handling.
    ///
    /// - Parameters:
    ///   - endpoint: the API Method (i.e. the endpoint for the request). Example: "/sources"
    ///   - parameters: a dictionary with all the parameters for the request
    ///   - completionHandler: a closure to be executed when the request is completed, which provides two (mutually exclusive) parameters: a dictionary containing JSON data, or a RequestError
    fileprivate func taskForGetMethod(endpoint: String, parameters: [String: String]? = nil, completionHandler: @escaping ([String:AnyObject]?, Error?) -> Void) {
        
        // Build the URL for the request from parameters and method
        let apiUrl = parseApiUrl(apiMethod: endpoint, parameters: parameters)
        
        // Build the request (defaults to "GET" request)
        let request = URLRequest(url: apiUrl)
        
        // Make the request
        let task = URLSession.shared.dataTask(with: request, completionHandler: { (data, response, error) in
            
            // GUARD: if there was an error with the data Task, just pass the error to the completion handler and return early
            guard (error == nil) else {
                completionHandler(nil, error)
                return
            }
            
            // GUARD: if the the response is invalid or the request was not successful (the response code is not 2xx),
            // pass a custom error to the completion handler and return early
            if let statusCode = (response as? HTTPURLResponse)?.statusCode {
                guard statusCode >= 200 && statusCode <= 299 else {
                    completionHandler(nil, RequestError.unsuccessfulRequest(statusCode: statusCode))
                    return
                }
            } else {
                completionHandler(nil, RequestError.invalidResponse)
                return
            }
            
            
            // GUARD: If no data was returned, send
            guard let data = data else {
                completionHandler(nil, RequestError.noData)
                return
            }
            
            // Try to parse data. If this fails, pass the error to
            var parsedResult: [String:AnyObject]!
            do {
                parsedResult = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as! [String:AnyObject]
            } catch {
                completionHandler(nil, RequestError.failedToParseJSON)
                return
            }
            
            completionHandler(parsedResult, nil)
        })
        
        // Start the request
        task.resume()
    }
    
    /// Builds the URL for a GET Request with the API method and eventually some parameters
    ///
    /// - Parameters:
    ///   - apiMethod: the API Method (i.e. the endpoint for the request)
    ///   - parameters: a dictionary with all the parameters for the request
    /// - Returns: the built URL for the request
    fileprivate func parseApiUrl(apiMethod: String, parameters: [String : String]?) -> URL {
        
        var components = URLComponents()
        components.scheme = Constants.URL.scheme
        components.host = Constants.URL.host
        components.path = Constants.URL.path + apiMethod
        
        if let parameters = parameters {
            components.queryItems = [URLQueryItem]()
            for (key, value) in parameters {
                let queryItem = URLQueryItem(name: key, value: value)
                components.queryItems!.append(queryItem)
            }
        }
        
        return components.url!
    }
    
    // MARK: - Errors
    
    /// Possible errors which can arise during the HTTP Request
    ///
    /// - invalidResponse: when the web service does not return a valid HTTP Response
    /// - unsuccessfulRequest: when the request is not successful (i.e. the response is not 2xx). The status code of the unsuccessful request is provided as argument
    /// - noData: when no Data is returned by the web service
    /// - failedToParseJSON: when the parsing from Data to JSON fails
    public enum RequestError: Error {
        case invalidResponse
        case unsuccessfulRequest(statusCode: Int)
        case noData
        case failedToParseJSON
    }
    
    /// Possible errors which can be reported by API's response
    ///
    /// - apiError: when the API reports an error due to incorrect parameters or other internal errors
    /// - noSource: when the API does not specify the source id for the provided articles
    /// - noSources: when the API does not include a "sources" object in the response after a request to the Sources endpoint
    /// - noArticles: when the API does not include an "articles" object in the response after a request to the Articles endpoint
    /// - emptySources: when the API responds with an empty array of Sources
    /// - emptyArticles: when the API responds with an empty array of Articles
    /// - wrongSourceFormat: when the API provides data in wrong format for Sources
    /// - wrongArticleFormat: when the API provides data in wrong format for Articles
    public enum ResponseError: Error {
        case networkError(message: String?)
        case apiError(message: String?)
        case noSource
        case noSources
        case noArticles
        case emptySources
        case emptyArticles
        case wrongSourceFormat
    }
}


NewsAPIClient/Classes/Constants.swift
//
//  Constants.swift
//  NewsAPIClient
//
//  Created by Andrea Marcolin (Twitter @andreamarcolin5) on 14/12/16.
//  Some rights reserved: http://opensource.org/licenses/MIT
//

import Foundation

extension NewsAPIClient {
    
    /// All the constant Strings used to interact with the API, such as URL components, endpoints and JSON keys
    struct Constants {
        
        /// URL components, such as scheme, host and path, to reach the API
        struct URL {
            static let scheme = "https"
            static let host = "newsapi.org"
            static let path = "/v1"
        }
        
        /// API endpoints
        struct Endpoints {
            static let articles = "/articles"
            static let sources = "/sources"
        }
        
        /// JSON request and response keys
        struct JSONKeys {
            
            // Articles JSON keys
            static let source = "source"
            static let apiKey = "apiKey"
            static let sortBy = "sortBy"
            static let articles = "articles"
            static let author = "author"
            static let title = "title"
            static let urlToImage = "urlToImage"
            static let publishedAt = "publishedAt"
            
            // Sources JSON keys
            static let sources = "sources"
            static let name = "name"
            static let category = "category"
            static let language = "language"
            static let country = "country"
            static let urlToLogos = "urlsToLogos"
            static let sortBysAvailable = "sortBysAvailable"
            
            // Common JSON keys
            static let status = "status"
            static let id = "id"
            static let description = "description"
            static let url = "url"
            static let message = "message"
        }
        
        /// Status values
        struct Status {
            static let ok = "ok"
            static let error = "error"
        }
        
        struct Utilities {
            static let dateFormats = ["yyyy-MM-dd'T'HH:mm:ss'Z'", "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", "yyyy-MM-dd'T'HH:mm:ss'+'hh:mm"]
        }
    }
}
